### DISCUSSION TITLE
Circuit Design for variable sized inputs

### DISCUSSION THREAD

**Discussion by shreyas-londhe** [Consensys/gnark#335] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/62744899/178117819-efa7d752-1a79-4569-8718-1ef53142a2da.png)

I came across this issue when I compiled the circuit with more number of variables. But then when I passed less number of variables, it gave me this error. Not sure why I'm getting this error. My circuit logic demands that the inputs will be of variable length everytime. is there a workaround? Any help will be appreciated, thanks!

**TLDR** - How to design circuit and compile for variable sized inputs.

**Comment by gbotrel** [comment, marked as answer]
hi -- not sure what you mean by "variable sized inputs". Constraint systems (i.e circuits) have fixed-size input at compile time. Do you have a code sample you can share?

**Reply by shreyas-londhe** [reply]
```go
type Data struct {
    hash            frontend.Variable      `gnark:public`
    preimage     []frontend.Variable
}

func (circuit *Data) Define() {
    for i := 0; i < len(preimage); i++ {
        hash := mimc(preimage[i])
    }
}
```
So suppose in this, the preimage is going to be a variable length array, so suppose I compile and setup the circuit with 10 array variables, the next time I run the circuit with 5 array variables using the previously generated proving and verifying keys. I get the above error. How do I overcome this issue?

I enquired about the same error in a hitter lobby and got the following response:

![image](https://user-images.githubusercontent.com/62744899/178312936-89ced776-f769-4475-bfcd-a4528df29029.png)

p.s - this is just a representation of my problem. not the actual thing. I'm verifying EdDSA signatures and checking merkle proofs in the actual loop.

**Reply by gbotrel** [reply]
Yes so; short version, you can't do that. 
Usual solution; have multiple circuits for multiple sizes. (that's what rollup project on Ethereum do; also, a small circuit will be way more efficient than a very large one)
You can have a single circuit of a large size, and "zero" part of the witness (such that it doesn't change the circuit valid inputs logic), but in some instances, you'll have to be careful in how this is designed to not allow invalid inputs (in your context) to generate a valid proof.

**Reply by shreyas-londhe** [reply]
Is it possible to combine multiple proofs into a single proof? I understand that the proof has - a ([2]bigint), b ([2][2]bigint), c ([2]bigint) and publicInput ([fixed]bigint). so If I get multiple proofs, is it possible to combine all these proofs into a single proof so that I can verify a single proof in the Solidity verifier. I know that the publicInputs cannot be combined and just need to be concatenated, but what about the - a, b, and c components, can I combine multiple a's, b's, and c's?

**Reply by shreyas-londhe** [reply]
I just read about something called Recursive proofs, which I can use to generate an aggregate proof. is this the right way to go and if possible can you please shoot me some references where I can see how it is implemented?

I am using bn254 for my primary circuit and in the gnark playground it is using BW6_761 / BLS12_377 pair to get recursive proof, can you please explain this a bit and why we are using a pair for this. Thank you!

**Reply by ivokub** [reply]
Have a look at the Groth16 verifier gadget https://github.com/ConsenSys/gnark/blob/develop/std/groth16_bls12377. The only caveat with the gadget is that it is only possible to recursively verify proofs on a 2-chain of curves. This means that if you generate the proof on BLS12-377 (or BLS24-315), then it can only be verified on BW6-761 (respectively BW6-633).
This has to do with the compatibility of the base and scalar fields of the corresponding curves. If the fields are not compatible then we would need to use some alternative methods, which are either not implemented yet or very inefficient. See https://medium.com/delendum/field-selection-for-recursive-snarks-726ad56c3a3c for longer explanation.

There is additional way to compress the recursive checks using [SnarkPack](https://eprint.iacr.org/2021/529), but gnark doesn't yet support it out-of-box.

**Reply by shreyas-londhe** [reply]
Thanks for such a detailed answer! 

I'm using the curve BN254 for my circuit and even the hash function and everything uses BN254, so does that mean that I'll have to shift to BLS12-377 in order to use recursive proofs or is there some other workaround?

**Reply by ivokub** [reply]
Short-term there isn't a better solution than to change the underlying curve. Currently the only viable alternative would be to verify the Groth16 proof using non-native arithmetic, but it is really-really inefficient. The ideas is that we can emulate any field on any scalar field (i.e. any curve), but this needs a lot of binary decompositions and this adds up to a lot of constraints, making proof generation quite time-consuming. I haven't myself tried it out, but theoretically it should be doable (maybe you would have to reimplement Groth16 verifier gadget to take as inputs `nonnative.Element` values, not `frontend.Variable` values). Check the field emulation gadget in `std/math/nonnative`. It is currently only available in `develop` branch.

**Reply by ivokub** [reply]
Long-term we are looking into instantiating proof systems on any field, but it is still yet very early.

**Reply by shreyas-londhe** [reply]
Thank you so much for this answer. So I think the only way for me to move forward is to set a max number of constraints and fill in dummy values.
